connection-limit/connection_limit_kern.c

```diff
@@ -273,17 +273,20 @@ int _xdp_limit_conn(struct xdp_md *ctx)
     if (data + sizeof(struct ethhdr)+ 1 > data_end)
         return XDP_PASS;
 
-    /* Check if its valid ip packet */
+    
     struct iphdr *iph = (struct iphdr *)(data + sizeof(struct ethhdr));
-    if (iph + 1 > data_end)
+    __u8 l4_offset = iph->ihl * 4; // ip header length
+
+    /* Check if its valid ip packet */
+    if ((unsigned char *)iph + l4_offset > data_end)
         return XDP_PASS;
 
     /* Ignore other than TCP packets */
     if (iph->protocol != IPPROTO_TCP)
         return XDP_PASS;
 
     /* Check if its valid tcp packet */
-    struct tcphdr *tcph = (struct tcphdr *)(iph + 1);
+    struct tcphdr *tcph = (struct tcphdr *)((unsigned char *)iph + l4_offset);
     if (tcph + 1 > data_end)
```

Comment:
TCP header is also a variable-length header, which is determined by Data Offset field. Apply the same IHL logic here.